lightning-block-sync/src/init.rs

```diff
@@ -0,0 +1,427 @@
+}
+
+/// A wrapper to make a cache read-only.
+///
+/// Used to prevent losing headers that may be needed to disconnect blocks common to more than one
+/// listener.
+struct ReadOnlyCache<'a, C: Cache>(&'a mut C);
+
+impl<'a, C: Cache> Cache for ReadOnlyCache<'a, C> {
+	fn look_up(&self, block_hash: &BlockHash) -> Option<&ValidatedBlockHeader> {
+		self.0.look_up(block_hash)
+	}
+
+	fn block_connected(&mut self, _block_hash: BlockHash, _block_header: ValidatedBlockHeader) {
+		unreachable!()
+	}
+
+	fn block_disconnected(&mut self, _block_hash: &BlockHash) -> Option<ValidatedBlockHeader> {
+		None
+	}
+}
+
+/// Wrapper for supporting dynamically sized chain listeners.
+struct DynamicChainListener<'a>(&'a mut dyn ChainListener);
+
+impl<'a> ChainListener for DynamicChainListener<'a> {
+	fn block_connected(&self, _block: &Block, _height: u32) {
+		unreachable!()
+	}
+
+	fn block_disconnected(&self, header: &BlockHeader, height: u32) {
+		self.0.block_disconnected(header, height)
+	}
+}
+
+/// A set of dynamically sized chain listeners, each paired with a starting block height.
+struct ChainListenerSet<'a>(Vec<(u32, &'a mut dyn ChainListener)>);
+
+impl<'a> ChainListener for ChainListenerSet<'a> {
+	fn block_connected(&self, block: &Block, height: u32) {
+		for (starting_height, chain_listener) in self.0.iter() {
+			if height > *starting_height {
+				chain_listener.block_connected(block, height);
+			}
+		}
+	}
+
+	fn block_disconnected(&self, _header: &BlockHeader, _height: u32) {
+		unreachable!()
+	}
+}
+
+impl<S, B: Deref, F: Deref, L: Deref> ChainListener for (RefCell<ChannelMonitor<S>>, B, F, L)
+where
+	S: keysinterface::Sign,
+	B::Target: chaininterface::BroadcasterInterface,
+	F::Target: chaininterface::FeeEstimator,
+	L::Target: logger::Logger,
+{
+	fn block_connected(&self, block: &Block, height: u32) {
+		let txdata: Vec<_> = block.txdata.iter().enumerate().collect();
+		self.0.borrow_mut().block_connected(&block.header, &txdata, height, &*self.1, &*self.2, &*self.3);
+	}
+
+	fn block_disconnected(&self, header: &BlockHeader, height: u32) {
+		self.0.borrow_mut().block_disconnected(header, height, &*self.1, &*self.2, &*self.3);
+	}
+}
+
+impl<S, M: Deref, B: Deref, K: Deref, F: Deref, L: Deref> ChainListener for &ChannelManager<S, M, B, K, F, L>
+where
+	S: keysinterface::Sign,
+	M::Target: chain::Watch<S>,
+	B::Target: chaininterface::BroadcasterInterface,
+	K::Target: keysinterface::KeysInterface<Signer = S>,
+	F::Target: chaininterface::FeeEstimator,
+	L::Target: logger::Logger,
+{
+	fn block_connected(&self, block: &Block, height: u32) {
+		let txdata: Vec<_> = block.txdata.iter().enumerate().collect();
+		ChannelManager::block_connected(self, &block.header, &txdata, height);
+	}
+
+	fn block_disconnected(&self, header: &BlockHeader, _height: u32) {
+		ChannelManager::block_disconnected(self, header);
+	}
+}
+
+impl<S, C: Deref, T: Deref, F: Deref, L: Deref, P: Deref> ChainListener for &ChainMonitor<S, C, T, F, L, P>
```

Comment:
Implemented for `Z`? When attempting this for these, I get:

```
error[E0119]: conflicting implementations of trait `ChainListener`:`
```

Were you able to get something to compile?

Reply:
More specifically, after temporarily removing the other two implementations used in the example, I get this error:

```term
error[E0119]: conflicting implementations of trait `ChainListener`:
   --> lightning-block-sync/src/init.rs:246:1
    |
226 | / impl<T, S, M: Deref, B: Deref, K: Deref, F: Deref, L: Deref> ChainListener for T
227 | | where
228 | |     T: Deref<Target = ChannelManager<S, M, B, K, F, L>>,
229 | |     S: keysinterface::Sign,
...   |
243 | |     }
244 | | }
    | |_- first implementation here
245 | 
246 | / impl<T, S, C: Deref, B: Deref, F: Deref, L: Deref, P: Deref> ChainListener for T
247 | | where
248 | |     T: Deref<Target = ChainMonitor<S, C, B, F, L, P>>,
249 | |     S: keysinterface::Sign,
...   |
263 | |     }
264 | | }
    | |_^ conflicting implementation
```

Reply:
Ah. I didn't try to implement this suggestion but that makes sense - `Trait<A=B>` is the same type as `Trait<A=C>` (but `Trait<A>` is not the same as `Trait<B>`). You could, however, implement it for `ChainMonitor` and `ChannelManager` and then also for `Deref<Target=B: ChainListener>`.

Reply:
Nice! That avoids the conflict but note that when implementing for `Deref` I get this warning:

```
warning: trait objects without an explicit `dyn` are deprecated
   --> lightning-block-sync/src/init.rs:281:42
    |
281 | impl<T: ChainListener> ChainListener for Deref<Target = T> {
    |                                          ^^^^^^^^^^^^^^^^^ help: use `dyn`: `dyn Deref<Target = T>`
    |
    = note: `#[warn(bare_trait_objects)]` on by default
```

Which I think we want to avoid doing? If I introduce another type parameter for `Deref` that brings me back to having conflicting implementations.

Reply:
I think you just need another parameter: `impl<D: Deref, T: ChainListener> ChainListener for D where D::Target = T {`

Reply:
Yeah, that's what I was referring to though when I said it brings me back to having a conflicting implementation.

```
error[E0119]: conflicting implementations of trait `ChainListener` for type `lightning::ln::channelmanager::ChannelManager<_, _, _, _, _, _>`:
   --> lightning-block-sync/src/init.rs:281:1
    |
243 | / impl<S, M: Deref, B: Deref, K: Deref, F: Deref, L: Deref> ChainListener for ChannelManager<S, M, B, K, F, L>
244 | | where
245 | |     S: keysinterface::Sign,
246 | |     M::Target: chain::Watch<S>,
...   |
259 | |     }
260 | | }
    | |_- first implementation here
...
281 |   impl<T: ChainListener, U: Deref<Target = T>> ChainListener for U {
    |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `lightning::ln::channelmanager::ChannelManager<_, _, _, _, _, _>`
    |
    = note: upstream crates may add a new impl of trait `std::ops::Deref` for type `lightning::ln::channelmanager::ChannelManager<_, _, _, _, _, _>` in future versions

error: aborting due to previous error
```

Reply:
Hmmm... so if I'm reading the error correctly, I think we can move `ChainListener` to the `lightning` crate and define the relevant listeners there.

Reply:
Huh. Yes, I think that would fix the issue.

Reply:
More precisely, implementing `ChainListener` for types in the `lightning` crate (e.g., `ChannelManager`).